Add AnchoredFragment

mrBliss opened the pull request:
An `AnchoredFragment` is a `ChainFragment` that is anchored to some start
`Point`: we know the `Point` right before the first, leftmost block in the
fragment.

edsko commented on ouroboros-network/src/Ouroboros/Network/AnchoredFragment.hs:
```diff
@@ -0,0 +1,554 @@
+
+-- | An 'AnchoredFragment' is a 'ChainFragment' that is anchored to some start
+-- 'Point': we know the 'Point' right before the first, leftmost block in the
+-- fragment. The anchor point can be thought of as a left exclusive bound.
```
Markdown notation instead of Haddock notation?

edsko replied:
Uh, nope, my bad, got confused.

edsko commented on ouroboros-network/src/Ouroboros/Network/AnchoredFragment.hs:
```diff
@@ -0,0 +1,554 @@
+import qualified Ouroboros.Network.ChainFragment as CF
+
+
+-- | An 'AnchoredFragment' is a 'ChainFragment' that is anchored to some start
```
"start point" sounds a bit awkward

edsko commented on ouroboros-network/src/Ouroboros/Network/AnchoredFragment.hs:
```diff
@@ -0,0 +1,554 @@
+--
+-- The fact that it is an /exclusive/ bound is particularly convenient when
+-- dealing with Genesis, which is the start of the chain, but not an actual
+-- block, but there /is/ a 'Point' that refers to Genesis
```
slightly awkward sentence construction here... "which is.. but not an... but there is..".

edsko commented on ouroboros-network/src/Ouroboros/Network/AnchoredFragment.hs:
```diff
@@ -0,0 +1,554 @@
+-- We return:
+--
+-- * @p@ anchored at @a@: the common prefix shared between the fragments @c1@
+--   and @c2@. We will call the head point of this prefix @a'@.
```
Slightly confusing why we call this `a'`, let's clarify that upfront.

edsko commented on ouroboros-network/src/Ouroboros/Network/AnchoredFragment.hs:
```diff
@@ -0,0 +1,554 @@
+-- >               └───┘     ├───┤
+-- >                         │ e'│
+-- >                         └───┘
+-- >      (p,       s1,       s2)
```
Soooo pretty :heart_eyes: :)

mrBliss commented:
> Nice work :) API can probably be shrunk a bit but let's hold off until we understand better how exactly this fits in.

Thanks!

I have addressed your comments and force-pushed.
